psi4/driver/p4util/exceptions.py

```diff
@@ -148,6 +150,44 @@ def __init__(self, eqn_description, iteration, wfn, e_conv, d_conv):
         self.wfn = wfn
 
 
+class TDSCFConvergenceError(ConvergenceError):
+    """Error called for problems with TDSCF iterations.
+
+    Parameters
+    ----------
+    wfn : psi4.core.Wavefunction
+        Wavefunction at time of exception
+    what : str
+        What we were trying to solve for (singlets/triplets, irrep) when we failed to converge
+    stats : Dict
+        Dictionary of convergence statistics of last iteration.
+        Keys are:
+
+          count : int, iteration number
+          res_norm : np.ndarray (nroots, ), the norm of residual vector for each roots
+          val : np.ndarray (nroots, ), the eigenvalue corresponding to each root
+          delta_val : np.ndarray (nroots, ), the change in eigenvalue from the last iteration to this ones
+          collapse : bool, if a subspace collapse was performed
+          product_count : int, the running total of product evaluations that was performed
+          done : bool, if all roots were converged
+    """
+
+    def __init__(self, iteration, wfn, what, stats):
+        # prepare message, including excitation energies and residual norm
+        conv_info = "==> Convergence statistics from last iteration <==\n\n"
+        conv_info += "Excitation Energy".center(21) + f" {'D[value]':^15}" + "|R|".center(11) + "\n"
+        conv_info += f"{'-':->20} {'-':->15} {'-':->15}\n"
+        for e, diff, r_norm in zip(stats["val"], stats["delta_val"], stats["res_norm"]):
+            conv_info += f"      {e:.6f}         {diff:-11.5e}    {r_norm:12.5e}\n"
+        ConvergenceError.__init__(self,
+                                  eqn_description=f"""TDSCF solver ({what})""",
+                                  iteration=iteration,
+                                  additional_info=conv_info)
+        self.wfn = wfn
+        self.what = what
+        self.stats = stats
+
```

Comment:
This is awesome!  Great to have a clear picture of how things stand and how far from convergence the calculation is